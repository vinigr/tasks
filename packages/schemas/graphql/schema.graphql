"""The root of all... queries"""
type Query {
  """Fetches an object given its ID"""
  node(
    """The ID of an object"""
    id: ID!
  ): Node
  me: User
  user(id: ID!): User
  task(id: ID!): Task
  tasks(after: String, first: Int, before: String, last: Int, search: String): TaskConnection
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

"""User Data"""
type User implements Node {
  """The ID of an object"""
  id: ID!
  _id: String
  name: String
  email: String
  active: Boolean
}

"""Task Data"""
type Task implements Node {
  """The ID of an object"""
  id: ID!
  _id: String
  title: String!
  description: String
  author: User
  createdAt: String
  updatedAt: String
}

"""A connection to a list of items."""
type TaskConnection {
  """Number of items in this connection"""
  count: Int!

  """
  A count of the total number of objects in this connection, ignoring pagination.
    This allows a client to fetch the first five objects by passing "5" as the
    argument to "first", then fetch the total count so it could display "5 of 83",
    for example.
  """
  totalCount: Int!

  """Offset from start"""
  startCursorOffset: Int!

  """Offset till end"""
  endCursorOffset: Int!

  """Information to aid in pagination."""
  pageInfo: PageInfoExtended!

  """A list of edges."""
  edges: [TaskEdge]!
}

"""Information about pagination in a connection."""
type PageInfoExtended {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""An edge in a connection."""
type TaskEdge {
  """The item at the end of the edge"""
  node: Task

  """A cursor for use in pagination"""
  cursor: String!
}

type Mutation {
  UserRegisterWithEmail(input: UserRegisterWithEmailInput!): UserRegisterWithEmailPayload
  UserLoginWithEmail(input: UserLoginWithEmailInput!): UserLoginWithEmailPayload
  TaskAdd(input: TaskAddInput!): TaskAddPayload
  TaskUpdate(input: TaskUpdateInput!): TaskUpdatePayload
  TaskDelete(input: TaskDeleteInput!): TaskDeletePayload
}

type UserRegisterWithEmailPayload {
  token: String
  error: String
  clientMutationId: String
}

input UserRegisterWithEmailInput {
  name: String!
  email: String!
  password: String!
  clientMutationId: String
}

type UserLoginWithEmailPayload {
  token: String
  error: String
  clientMutationId: String
}

input UserLoginWithEmailInput {
  email: String!
  password: String!
  clientMutationId: String
}

type TaskAddPayload {
  task: TaskEdge
  error: String
  clientMutationId: String
}

input TaskAddInput {
  title: String!
  description: String
  clientMutationId: String
}

type TaskUpdatePayload {
  task: TaskEdge
  error: String
  clientMutationId: String
}

input TaskUpdateInput {
  id: ID!
  title: String!
  description: String
  clientMutationId: String
}

type TaskDeletePayload {
  id: ID
  error: String
  clientMutationId: String
}

input TaskDeleteInput {
  id: ID!
  clientMutationId: String
}

type Subscription {
  TaskAdded: TaskAddedPayload
  TaskUpdated: TaskUpdatedPayload
  TaskDeleted: TaskDeletedPayload
}

type TaskAddedPayload {
  taskEdge: TaskEdge
}

type TaskUpdatedPayload {
  taskEdge: TaskEdge
}

type TaskDeletedPayload {
  id: ID
}
