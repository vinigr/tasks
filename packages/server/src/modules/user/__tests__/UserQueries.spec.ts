import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { schema } from '../../../schema';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
  sanitizeTestObject,
} from '../../../../test/helpers';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('UserType queries', () => {
  it('should query an user', async () => {
    const user = await createUser();

    // language=GraphQL
    const query = `
      query Q($id: ID!) {
        user: node(id: $id) {
          id
          ... on User {
            id
            name
            email
          }
        }
      }
    `;

    const variables = {
      id: toGlobalId('User', user._id),
    };
    const rootValue = {};
    const context = await getContext();
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });
});
