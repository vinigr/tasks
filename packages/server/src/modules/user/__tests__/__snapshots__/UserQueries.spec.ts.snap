// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserType queries should query an user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "email": "EMPTY",
      "id": "FROZEN-ID",
      "name": "Normal user 1",
    },
  },
}
`;
